### PR TITLE
docs(skill): add object-model traps reference

### PR DESCRIPTION
**Pull request by Ryujiyasu**
Adds `references/object-model-traps.md` and its dispatch entry in `SKILL.md`, as
agreed in #679.

## Summary

Six entries covering Excel runtime behaviour that xlflow structurally cannot
diagnose, because the answer depends on the data rather than on the source:

- approximate lookup over unsorted data
- criteria comparisons not being symmetric (`<>` is not the negation of `=`)
- insert and delete moving absolute references
- a partial shift dividing formulas and merges oppositely
- sorting ordering by kind before value, and the `Header` default
- the same name being two different functions (`Round`, `Trim`)

Each is written as **Trigger / Risk / Safe rule / Proof / Scope-provenance**, per
the shape you set out. The file ends with a short safety list, following
`formulas.md`.

## Following the constraints from the issue

**Observation is kept separate from contract.** For approximate `MATCH` on
unsorted data, the Safe rule is "sort first or pass exact matching; do not rely
on any particular wrong answer". The specific result — position 3, the cell
holding 20 — sits in the Scope section, labelled as one implementation's
behaviour on one environment, with an explicit note not to treat it as a
guaranteed semantic. The same split is applied to the merge half of the partial
shift entry, and to the criteria counts.

**Statically detectable cases are out.** Omitted approximate-match arguments and
unsupported `WorksheetFunction` members are not in this file. They belong in
diagnostics, as you said.

**Nothing binds the VBE oracle.** No fixture is added and nothing here claims
compile-equivalent evidence. Provenance is carried per entry, and the probes
below are supplied for review rather than as corpus.

**Where Excel was genuinely unstable, the case is absent.** A descending
approximate match over shuffled input was not reproducible across runs, so it is
in neither the reference nor the probes, and the reference says so.

## Verification

Every number quoted in the reference was re-run over COM against Excel 16
(Windows 11, ja-JP) while preparing this PR. That caught one error in my own
draft: I had written that a partial `xlShiftDown` grows `SUM(A1:C3)` to
`SUM(A1:C4)`. It does not — the range reaches past the shifted band, so it is
left exactly as it was. The reference and the probes now say that, with a
contrast against the whole-row insert in Probe 3, where the range *does* resize.

Output of the run:

```
p1 unsorted Match(35)=3 cell=20
p1 sorted   Match(35)=3 cell=30
p2 =20 -> 3    <>20 -> 2    >=20 -> 1
p3 C2==$A$3   C3==SUM(A2:A4)
p4 merged before: B5:C5
p4 formula: =SUM(A1:C3)   C3: kept   merged after: B6
p5 order: [5][header][text][True][]
p6 WF.Round(2.5,0)=3  WF.Round(3.5,0)=4  WF.Trim=[a b]
```

`p6`'s VBA-side halves (`Round(2.5, 0)` → 2, `Trim` preserving inner runs) are
documented language behaviour and are asserted by Probe 6 rather than by that
COM run, which can only reach `WorksheetFunction`.

## Probes

Self-contained — each writes its own input, so none needs a prepared workbook.
Happy to move these into the repository in whatever form you prefer, or to leave
them here if the reference's Proof sections are enough.

<details>
<summary>Six probe procedures</summary>

## Probe 1 — approximate lookup over unsorted data

```vba
Public Sub ProbeUnsortedMatch()
    Dim sheet As Worksheet
    Set sheet = ThisWorkbook.Worksheets.Add
    sheet.Range("A1:A5").Value = Application.Transpose(Array(10, 50, 20, 40, 30))

    Debug.Print "Match(35, unsorted, 1): " & _
        WorksheetFunction.Match(35, sheet.Range("A1:A5"), 1) & " | 3"
    Debug.Print "cell at that position:  " & _
        sheet.Range("A1:A5").Cells(WorksheetFunction.Match(35, sheet.Range("A1:A5"), 1)).Value & " | 20"

    ' The same data sorted, where approximate matching is defined.
    sheet.Range("A1:A5").Sort Key1:=sheet.Range("A1"), Order1:=xlAscending, Header:=xlNo
    Debug.Print "Match(35, sorted, 1):   " & _
        WorksheetFunction.Match(35, sheet.Range("A1:A5"), 1) & " | 3"
    Debug.Print "cell at that position:  " & _
        sheet.Range("A1:A5").Cells(WorksheetFunction.Match(35, sheet.Range("A1:A5"), 1)).Value & " | 30"

    Application.DisplayAlerts = False
    sheet.Delete
    Application.DisplayAlerts = True
End Sub
```

Both calls return 3. Only the second one means the nearest value at or below 35.

## Probe 2 — criteria comparisons are not symmetric

```vba
Public Sub ProbeCriteriaAsymmetry()
    Dim sheet As Worksheet
    Set sheet = ThisWorkbook.Worksheets.Add
    sheet.Range("A1").NumberFormat = "@"
    sheet.Range("A1").Value = "20"        ' text
    sheet.Range("A2").Value = 20          ' number
    sheet.Range("A3").NumberFormat = "@"
    sheet.Range("A3").Value = "20.0"      ' text

    Debug.Print "CountIf =20:  " & WorksheetFunction.CountIf(sheet.Range("A1:A3"), "=20") & " | 3"
    Debug.Print "CountIf <>20: " & WorksheetFunction.CountIf(sheet.Range("A1:A3"), "<>20") & " | 2"
    Debug.Print "CountIf >=20: " & WorksheetFunction.CountIf(sheet.Range("A1:A3"), ">=20") & " | 1"
    Debug.Print "cells in range: " & sheet.Range("A1:A3").Count & " | 3"

    Application.DisplayAlerts = False
    sheet.Delete
    Application.DisplayAlerts = True
End Sub
```

Three cells, five matches across `=` and `<>`.

## Probe 3 — insert moves an absolute reference

```vba
Public Sub ProbeAbsoluteShift()
    Dim sheet As Worksheet
    Set sheet = ThisWorkbook.Worksheets.Add
    sheet.Range("A1:A3").Value = Application.Transpose(Array(1, 2, 3))
    sheet.Range("C1").Formula = "=$A$2"
    sheet.Range("C2").Formula = "=SUM(A1:A3)"

    Debug.Print "before, C1: " & sheet.Range("C1").Formula & " | =$A$2"
    Debug.Print "before, C2: " & sheet.Range("C2").Formula & " | =SUM(A1:A3)"

    sheet.Rows(1).Insert
    Debug.Print "after insert, C1: " & sheet.Range("C2").Formula & " | =$A$3"
    Debug.Print "after insert, C2: " & sheet.Range("C3").Formula & " | =SUM(A2:A4)"

    Application.DisplayAlerts = False
    sheet.Delete
    Application.DisplayAlerts = True
End Sub
```

The formulas moved down with the insert, so they are read from `C2`/`C3`
afterwards. `$A$2` became `$A$3`; the SUM range resized rather than shifted.

## Probe 4 — a partial shift divides formulas and merges oppositely

```vba
Public Sub ProbePartialShift()
    Dim sheet As Worksheet
    Set sheet = ThisWorkbook.Worksheets.Add
    sheet.Range("A1:C3").Value = 1
    sheet.Range("E1").Formula = "=SUM(A1:C3)"
    sheet.Range("B3").Value = "kept"
    sheet.Range("C3").Value = "kept"
    sheet.Range("B5:C5").Merge

    Debug.Print "before, merged: " & (sheet.Range("B5").MergeArea.Address(False, False)) & " | B5:C5"

    sheet.Range("B2").Insert Shift:=xlShiftDown

    Debug.Print "formula reaching past the band: " & sheet.Range("E1").Formula & " | =SUM(A1:C3)"
    Debug.Print "C3 untouched by the band:      " & sheet.Range("C3").Value & " | kept"
    Debug.Print "merge half covered:            " & sheet.Range("B6").MergeArea.Address(False, False) & " | B6 (broken)"

    Application.DisplayAlerts = False
    sheet.Delete
    Application.DisplayAlerts = True
End Sub
```

The formula was left exactly as it was, because `A1:C3` reaches past the shifted
band; the merge the same band only half covered was taken apart. Same
containment test, opposite conclusions.

(Contrast with Probe 3, where a whole-row insert did resize `SUM(A1:A3)` to
`SUM(A2:A4)`. Whether the range moves depends on whether the edit spans it.)

## Probe 5 — sorting orders by kind before value

```vba
Public Sub ProbeSortByKind()
    Dim sheet As Worksheet
    Set sheet = ThisWorkbook.Worksheets.Add
    sheet.Range("A1").Value = "header"
    sheet.Range("A2").Value = True
    sheet.Range("A3").Value = "text"
    sheet.Range("A4").Value = 5
    sheet.Range("A5").Value = ""

    ' Header omitted, which defaults to xlNo.
    sheet.Range("A1:A5").Sort Key1:=sheet.Range("A1"), Order1:=xlAscending

    Dim at As Long
    For at = 1 To 5
        Debug.Print "row " & at & ": " & TypeName(sheet.Cells(at, 1).Value) & " " & _
            CStr(sheet.Cells(at, 1).Value)
    Next
    Debug.Print "expected: number 5, then text (header, text), then Boolean True, then blank"

    Application.DisplayAlerts = False
    sheet.Delete
    Application.DisplayAlerts = True
End Sub
```

The header sorted in with the data because `Header` was omitted. Numbers precede
text, which precedes Booleans; the blank stays last.

## Probe 6 — the same name is two different functions

```vba
Public Sub ProbeSameNameTwoFunctions()
    Debug.Print "VBA Round(2.5, 0):              " & Round(2.5, 0) & " | 2"
    Debug.Print "VBA Round(3.5, 0):              " & Round(3.5, 0) & " | 4"
    Debug.Print "WorksheetFunction.Round(2.5,0): " & WorksheetFunction.Round(2.5, 0) & " | 3"
    Debug.Print "WorksheetFunction.Round(3.5,0): " & WorksheetFunction.Round(3.5, 0) & " | 4"

    Debug.Print "VBA Trim:                       [" & Trim("  a   b  ") & "] | [a   b]"
    Debug.Print "WorksheetFunction.Trim:         [" & WorksheetFunction.Trim("  a   b  ") & "] | [a b]"
End Sub
```

VBA rounds half to even, `WorksheetFunction` rounds half away from zero. `3.5`
agrees under both rules, which is why the boundary case must be `2.5`.

## What these do not establish

Probe 1's unsorted result is one implementation's probe path, not a contract.
The reference states it as an observation for that reason. A descending
approximate match over shuffled input was not stable across runs and is
deliberately absent from both the reference and these probes.

</details>

## Summary by CodeRabbit

* **Documentation**
  * Expanded guidance for Excel runtime edge cases, including approximate lookups, criteria comparisons, structural edits, partial shifts, merged cells, mixed-type sorting, and rounding or trimming behavior.
  * Added verification steps, boundary-test recommendations, and consolidated safety rules.
  * Updated specialized reference routing to cover these scenarios.
* **Tests**
  * Updated installation checks to verify the new reference documentation and its dispatch link.

### CODE REVIEW

**Comment by Ryujiyasu** [comment]
Thanks — four of the five were right and are fixed in dbb4920. Re-ran everything
over COM before changing anything.

**`#N/A` (comment 1): correct, and my sentence was wrong.** A key below every
value does return it, and the `WorksheetFunction` call raises rather than
returning:

```
formula MATCH(5,A1:A5,1)  unsorted -> #N/A
formula MATCH(5,A1:A5,1)  sorted   -> #N/A
VLOOKUP(5,A1:A5,1)        sorted   -> #N/A
WorksheetFunction.Match(5, ...)    -> raises
formula MATCH(35,A1:A5,1) unsorted -> 3
```

Rewritten so the entry now says the out-of-range key is the loud case, and that
it is a key *inside* the range that comes back silently wrong. The probe shows
both side by side.

**Trim in the proof (comment 4): correct.** The entry documented both and proved
only one. It now asserts `Trim("  a   b  ")` against
`WorksheetFunction.Trim("  a   b  ")`, and notes that `Round(3.5, 0)` agrees
under both rules so only `2.5` distinguishes them.

**Dispatch row (comment 5): correct.** Partial shifts, merges under a shift, and
`WorksheetFunction` versus VBA were all covered by the file but not named in the
row, so those queries could not reach it. All are named now.

**Session recipes (comment 2): taken, in the shorter form.** The proof steps now
point at the session loop SKILL.md already lays out (`session start` → `push
--fast --session --no-save` → `test --session --no-save` → `save --session` →
`session stop`) rather than restating it per entry. The `formulas pull` recipes
already noted `save` first and are unchanged, since that command reads the saved
file rather than session state.

**Match count (comment 3): I do not think this one is right.** The counts are 3,
2 and 1, but `>=20` is a third, separate assertion — it is not part of the same
total. The point of the entry is that `=20` and `<>20` **overlap** instead of
partitioning: five matches between them over three cells. Adding `>=20` to that
sum measures nothing.

That said, the sentence was ambiguous enough to be read your way, so I have
made the claim explicit rather than leaving a total for the reader to
reconstruct, and the probe now prints the sum of just those two:

```
=20 is 3, <>20 is 2, together 5 over 3 cells
```

File: internal/agentskill/templates/xlflow/references/object-model-traps.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+— the default — against a range whose sort order you have not established in
+the same procedure.
+
+**Risk.** A key below every value in the range returns `#N/A` in a formula, and
+raises at the `WorksheetFunction` call — that case is at least loud. The
+dangerous one is a key inside the range: a position comes back with no error and
+no `#N/A`, and it is not the nearest value. The result may also be correct by
+accident on the data you tested with, and wrong on the data your user has.
+
+**Safe rule.** Either sort the range ascending immediately before the lookup, or
```
Comment (harumiWeb) [inline internal/agentskill/templates/xlflow/references/object-model-traps.md]: I think the overall guidance here is right, but I’d adjust the safe rule slightly.

“Sort the range ascending immediately before the lookup” can be read too literally by an agent. In particular, sorting only the lookup-key range for VLookup could break the row association with the return columns, and sorting the user’s data in place may itself be an unintended side effect.

I think the invariant we actually want is: approximate matching requires the required sort order to be established and verified; otherwise use exact matching. If sorting is necessary, the whole associated dataset must remain consistent.

I’d also avoid having the Proof assert a particular position from an unsorted approximate lookup, since the Scope section correctly says that result is not a contract. It would be safer to prove the corrected behavior instead: either exact matching works regardless of order, or the required ordering is established/validated before approximate matching.

File: internal/agentskill/templates/xlflow/references/object-model-traps.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+`SUM(A1:C3)` alone because that range reaches past column B — while a merge only
+half covered by the same band is broken. The formula half follows from documented
+reference adjustment; the merge half is an observation.
+
+## Sorting Orders By Kind Before Value
+
+**Trigger.** `Range.Sort` or `AutoFilter` sorting over a column that may hold more
+than one type, or any `Sort` call that omits `Header`.
+
+**Risk.** Two silent failures. Values do not interleave: all numbers come before
```
Comment (harumiWeb) [inline internal/agentskill/templates/xlflow/references/object-model-traps.md]: Small correctness issue here: the type ordering described in the Risk section is true for ascending sort, but descending reverses the type order.

Could we make the direction explicit? Something along the lines of:

`In ascending order, numbers precede text, which precedes Booleans; descending reverses the type order. Blanks remain at the bottom in both directions.`

The existing blank guidance is useful; I just want to avoid an agent generalizing the ascending ordering to descending sorts as well.

File: internal/agentskill/templates/xlflow/references/object-model-traps.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+formulas, and it is wrong here. `$A$2` is unchanged by a copy and becomes `$A$3`
+when a row is inserted above it. Formulas keep pointing at the value they meant,
+which is usually what you want — but if your code assumed the address text was
+stable, it is now reading a different cell.
+
+**Safe rule.** Never assume an address literal survives a structural edit. Re-read
+addresses after inserting or deleting, or refer to the data through a defined
+name or a table column, which are maintained across the edit.
+
+**Safe rule for ranges.** A range argument resizes rather than shifts. An insert
```
Comment (harumiWeb) [inline internal/agentskill/templates/xlflow/references/object-model-traps.md]: One wording tweak here: I think “A range argument resizes rather than shifts” is a little too absolute.

Structural edits can either shift or resize a referenced range depending on where the edit intersects it. The PR’s own whole-row probe demonstrates the shift case with `SUM(A1:A3)` becoming `SUM(A2:A4)`, while an insertion inside a referenced range demonstrates expansion.

Could we phrase this more generally, for example:

`A range reference may shift or resize depending on where the structural edit intersects it.`

Then the existing examples can explain the specific resize / `#REF!` cases.

**Review by harumiWeb — COMMENTED** [review]
Thanks for putting this together so quickly. The overall direction looks good, and I think the new reference fits the Skill structure well.

I especially appreciate that the PR keeps observed Excel behavior separate from documented/guaranteed semantics, and that statically detectable cases were kept out of this reference as discussed.

I left a few small inline comments around wording and proof guidance, mainly to make sure agents do not overgeneralize the examples or take an unsafe action too literally. I also noted one small regression-test update for the newly added reference.

Nothing changes the overall design here — once those points are addressed, this should be in good shape to merge. Thanks again for the thoughtful contribution.

Comments outside the diff range:

Could we also add `object-model-traps.md` to `referenceFiles` in `internal/agentskill/skill_test.go`, and ideally assert that the new dispatch link appears in the installed `SKILL.md`?

The current installer will already bundle the file because it embeds and walks the whole template tree, so this is not a functional bug today. However, `referenceFiles` appears to be the regression contract that verifies all required specialized references are actually present after installation, so the new reference should be covered there too.

**Comment by Ryujiyasu** [comment]
Thanks — all four are addressed in 9e4dd607. Re-measured over COM before
changing anything, and one of the measurements turned up something neither of us
had written down.

**Sort direction (comment 2).** You are right, and the reference was stating an
ascending-only fact as if it held generally. Measured over a column holding
`text, 5, TRUE, [blank], #DIV/0!, apple, 100, FALSE`:

```
ascending    5 | 100 | apple | text | FALSE | TRUE | #DIV/0! | [blank]
descending   #DIV/0! | TRUE | FALSE | text | apple | 100 | 5 | [blank]
```

So descending reverses the order of kinds and the values inside each kind, and
blanks stay at the bottom both ways, exactly as you said. The extra fact is that
**error values are part of the ordering** — last in ascending, first in
descending — while blanks sit outside it. The Risk section now names all four
kinds with the direction attached, the blanks rule says explicitly that they are
outside the ordering, and the Proof asks for both directions, since a test that
only covers ascending will not notice code that assumes the kinds stay put when
the direction flips. Both sequences are quoted in Scope as the observation.

**Shift or resize (comment 3).** Also right, and "resizes rather than shifts" was
me generalising from the one case I had probed. Measured across the four
intersections against `SUM(A1:A3)`:

```
whole row above        SUM(A2:A4)     shifted
whole row inside       SUM(A1:A4)     resized
whole row just past    SUM(A1:A3)     untouched
delete row inside      SUM(A1:A2)     resized
```

The sentence now reads "may shift or resize, depending on where the structural
edit meets it" and names each case, with the point that the adjustment is silent
either way carried over.

**Approximate lookup (comment 1).** Taken, including the part I had not thought
about: the rule as written would let an agent sort a `VLookup` key range on its
own and break its rows away from the return columns. The safe rule is now the
invariant — approximate matching is only defined when the required order is
established *and* verified, otherwise pass exact matching — followed by a
paragraph saying that sorting to satisfy it is a structural edit, must move the
whole associated dataset, and needs its own justification because it changes the
user's data.

The Proof no longer asserts the position an unsorted approximate lookup returns.
It asks for the corrected behaviour instead: either assert that exact matching
finds the key wherever it sits, or establish the required order, assert the order
itself, then assert the approximate result. It also says outright not to hold a
test against the observed position, with a pointer to the Scope section that
explains why.

**Regression test (outside the diff).** `object-model-traps.md` is in
`referenceFiles` now, and `TestInstallUsesProviderDefaultTarget` asserts
`[object-model-traps.md](references/object-model-traps.md)` alongside the
recovery and code-analysis links, so the dispatch row is covered as well as the
file. I have no Go toolchain on the machine I measured from, so that pair is
riding on CI rather than on a local run — worth a glance if it looks off.

The updated sort probe, printing both directions and the error value:

```vb
Public Sub ProbeSortDirection()
    Dim ws As Worksheet
    Set ws = ThisWorkbook.Worksheets(1)
    ws.Cells.Clear
    ws.Range("A1").Value = "text"
    ws.Range("A2").Value = 5
    ws.Range("A3").Value = True
    ws.Range("A5").Formula = "=1/0"
    ws.Range("A6").Value = "apple"
    ws.Range("A7").Value = 100
    ws.Range("A8").Value = False
    ws.Range("A1:A8").Sort Key1:=ws.Range("A1"), Order1:=xlAscending, Header:=xlNo
    Debug.Print "ascending  " & Report(ws)
    ' refill, then:
    ws.Range("A1:A8").Sort Key1:=ws.Range("A1"), Order1:=xlDescending, Header:=xlNo
    Debug.Print "descending " & Report(ws)
End Sub
```

**Comment by Ryujiyasu** [comment]
Withdrawing the caveat in my last comment: I have a Go toolchain now (1.27.0),
so the test change is verified locally rather than left to CI.

```
gofmt -l internal/agentskill/skill_test.go   (no output)
go vet ./internal/agentskill/                (clean)
go test ./internal/agentskill/               ok  1.533s
```

Worth knowing either way, since the workflow runs on 9e4dd607 are sitting at
`action_required` and will not tell you anything until you approve them.

**Comment by harumiWeb** [comment]
Thanks for addressing the review comments so thoroughly. The updated guidance and regression coverage look good to me.

Everything looks good now — thanks again for the thoughtful contribution!

**Comment by Ryujiyasu** [comment]
For completeness, now that the toolchain is in place: the whole branch is green locally, not just the package I touched — `go build ./...`, `go vet ./...` and `go test ./...` (53 packages, no failures) on 9e4dd607.
